[RISCV] Implement codegen for XAndesPerf lea instructions

tclin914 opened the pull request:
This patch add the patterns for generating XAndesPerf lea instructions.

The operation of LEA family instructions is:

  rd = rs1 + rs2 * (the number of bytes)

The variants with *.ze suffix are RV64 only and its operation is:

  rd = rs1 + ZE32(rs2[31:0]) * (the number of bytes)

tclin914 committed:
[RISCV] Implement codegen for XAndesPerf lea instructions

This patch add the patterns for generating XAndesPerf lea instructions.

The operation of LEA family instructions is:

  rd = rs1 + rs2 * (the number of bytes)

The variants with *.ze suffix are RV64 only and its operation is:

  rd = rs1 + ZE32(rs2[31:0]) * (the number of bytes)

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXAndes.td:
```diff
@@ -356,3 +356,28 @@ def NDS_LDGP  : NDSRVInstLDGP<0b011, "nds.ldgp">;
+
+let Predicates = [HasVendorXAndesPerf] in {
+class NDS_LEAPat<int shamt, RVInstR Inst>
+    : Pat<(add (XLenVT GPR:$rs1), (shl GPR:$rs2, (XLenVT shamt))),
```
You probably want to use `add_like_non_imm12` instead of `add`. That's what we do for sh1add/sh2add/sh3add from Zba. That will handle `or disjoint` and prevent using the instruction if an ADDI could be used.

tclin914 replied:
Thanks for pointing that out. I've updated the patterns for Zba to be multiclass/class so they can be reused.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXAndes.td:
```diff
@@ -356,3 +356,28 @@ def NDS_LDGP  : NDSRVInstLDGP<0b011, "nds.ldgp">;
+          (NDS_LEA_B_ZE GPR:$rs1, GPR:$rs2)>;
+
+class NDS_LEA_ZEPat<int shamt, RVInstR Inst>
+    : Pat<(add GPR:$rs1, (shl (zexti32 (XLenVT GPR:$rs2)), (XLenVT shamt))),
```
I don't think you want `zexti32` here. That will use this form if an AND is present or if the upper 32 bits are known to be 0. You probably only want the AND case. Use `(and GPR:$rs2, 0xFFFFFFFF)` like we do for sh1add/sh2add/sh3add.

wangpc-pp reviewed (commented):
IIUC, they are equivalent to Zba? Maybe we should reuse the base pattern?

tclin914 committed:
Make the patterns for Zba be multiclass/class that we can reuse it.

tclin914 committed:
Reuse the patterns from Zba for XAndesPerf lea instructions

tclin914 commented:
> IIUC, they are equivalent to Zba? Maybe we should reuse the base pattern?

Thanks for the reminder. I've update the patterns for Zba that we can reuse it.

tclin914 committed:
indent the body in multiclass by 2 spaces.

tclin914 commented on llvm/lib/Target/RISCV/RISCVInstrInfoZb.td:
```diff
@@ -659,30 +659,103 @@ def : Pat<(i32 (and GPR:$rs, 0xFFFF)), (PACK GPR:$rs, (XLenVT X0))>;
 
-let Predicates = [HasStdExtZba] in {
+multiclass ShxAddPat<int i, Instruction shxadd> {
+def : Pat<(XLenVT (add_like_non_imm12 (shl GPR:$rs1, (XLenVT i)), GPR:$rs2)),
```
Done.

tclin914 commented on llvm/lib/Target/RISCV/RISCVInstrInfoZb.td:
```diff
@@ -659,30 +659,103 @@ def : Pat<(i32 (and GPR:$rs, 0xFFFF)), (PACK GPR:$rs, (XLenVT X0))>;
 } // Predicates = [HasStdExtZba]
 
+multiclass ADD_UWPat<Instruction add_uw> {
+def : Pat<(i64 (add_like_non_imm12 (and GPR:$rs1, 0xFFFFFFFF), GPR:$rs2)),
```
Done.

tclin914 commented on llvm/lib/Target/RISCV/RISCVInstrInfoZb.td:
```diff
@@ -659,30 +659,103 @@ def : Pat<(i32 (and GPR:$rs, 0xFFFF)), (PACK GPR:$rs, (XLenVT X0))>;
+}
+
+multiclass ShxAdd_UWPat<int i, Instruction shxadd_uw> {
+def : Pat<(i64 (add_like_non_imm12 (shl (and GPR:$rs1, 0xFFFFFFFF), (i64 i)),
```
Done.

tclin914 commented on llvm/lib/Target/RISCV/RISCVInstrInfoZb.td:
```diff
@@ -659,30 +659,103 @@ def : Pat<(i32 (and GPR:$rs, 0xFFFF)), (PACK GPR:$rs, (XLenVT X0))>;
                   GPR:$r)>;
 
+let Predicates = [HasStdExtZba] in {
+
```
The bodies of the other `let` in this .td file don't have 2 spaces indentation. If I only make the change here, it won't have a consistent style throughout the file.

tclin914 committed:
Switch the position of rs1 and rs2 for lea instruction

In Zba, rs1 is index register and rs2 is base register. But for lea
instructions, rs1 is base register and rs2 is index register.

tclin914 commented:
Kindly ping.

topperc commented:
This is failing MC tests probably because you changed the order of $rs1 and $rs2 which broke the encoding/decoding.

tclin914 committed:
Change the order of rs1 and rs1 in ins rather than in argstr

tclin914 commented:
> This is failing MC tests probably because you changed the order of $rs1 and $rs2 which broke the encoding/decoding.

Fixed it.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXAndes.td:
```diff
@@ -125,19 +125,25 @@ class NDSRVInstBFO<bits<3> funct3, string opcodestr>
 class NDSRVInstLEA<bits<7> funct7, string opcodestr>
-    : NDSRVInstRR<funct7, opcodestr, "$rd, $rs2, $rs1">;
+    : RVInstR<funct7, 0b000, OPC_CUSTOM_2,
+              (outs GPR:$rd), (ins GPR:$rs2, GPR:$rs1),
```
Please add a comment about why the operands are reversed.

topperc reviewed (approved):
LGTM
